File: Basics/Tree Traversal/Tree.js
class TreeNode {
    constructor(value) {
      this.value = value;
      this.left = null;
      this.right = null;
    }
  }
  
  class BinaryTree {
    constructor() {
      this.root = null;
    }
  
    insert(value) {
      this.root = this._insert(this.root, value);
    }
  
    _insert(root, value) {
      if (root === null) {
        return new TreeNode(value);
      }
  
      if (value < root.value) {
        root.left = this._insert(root.left, value);
      } else {
        root.right = this._insert(root.right, value);
      }
  
      return root;
    }
  
    inOrderTraversal() {
      const result = [];
      this._inOrderTraversal(this.root, result);
      return result;
    }
  
    _inOrderTraversal(node, result) {
      if (node !== null) {
        this._inOrderTraversal(node.left, result);
        result.push(node.value);
        this._inOrderTraversal(node.right, result);
      }
    }
  
    preOrderTraversal() {
      const result = [];
      this._preOrderTraversal(this.root, result);
      return result;
    }
  
    _preOrderTraversal(node, result) {
      if (node !== null) {
        result.push(node.value);
        this._preOrderTraversal(node.left, result);
        this._preOrderTraversal(node.right, result);
      }
    }
  
    postOrderTraversal() {
      const result = [];
      this._postOrderTraversal(this.root, result);
      return result;
    }
  
    _postOrderTraversal(node, result) {
      if (node !== null) {
        this._postOrderTraversal(node.left, result);
        this._postOrderTraversal(node.right, result);
        result.push(node.value);
      }
    }
  }
  
  // Example usage:
  const myTree = new BinaryTree();
  myTree.insert(5);
  myTree.insert(3);
  myTree.insert(7);
  myTree.insert(2);
  myTree.insert(4);
  myTree.insert(6);
  myTree.insert(8);
  
  console.log("In-order traversal:", myTree.inOrderTraversal());
  console.log("Pre-order traversal:", myTree.preOrderTraversal());
  console.log("Post-order traversal:", myTree.postOrderTraversal());
  

File: Basics/Tree Traversal/Tree.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        self.root = self._insert(self.root, value)

    def _insert(self, root, value):
        if root is None:
            return TreeNode(value)

        if value < root.value:
            root.left = self._insert(root.left, value)
        else:
            root.right = self._insert(root.right, value)

        return root

    def in_order_traversal(self):
        result = []
        self._in_order_traversal(self.root, result)
        return result

    def _in_order_traversal(self, node, result):
        if node is not None:
            self._in_order_traversal(node.left, result)
            result.append(node.value)
            self._in_order_traversal(node.right, result)

    def pre_order_traversal(self):
        result = []
        self._pre_order_traversal(self.root, result)
        return result

    def _pre_order_traversal(self, node, result):
        if node is not None:
            result.append(node.value)
            self._pre_order_traversal(node.left, result)
            self._pre_order_traversal(node.right, result)

    def post_order_traversal(self):
        result = []
        self._post_order_traversal(self.root, result)
        return result

    def _post_order_traversal(self, node, result):
        if node is not None:
            self._post_order_traversal(node.left, result)
            self._post_order_traversal(node.right, result)
            result.append(node.value)

# Example usage:
my_tree = BinaryTree()
my_tree.insert(5)
my_tree.insert(3)
my_tree.insert(7)
my_tree.insert(2)
my_tree.insert(4)
my_tree.insert(6)
my_tree.insert(8)

print("In-order traversal:", my_tree.in_order_traversal())
print("Pre-order traversal:", my_tree.pre_order_traversal())
print("Post-order traversal:", my_tree.post_order_traversal())
